Workflow Invocation view improvements

ahmedhamidawan opened the pull request:
Following discussions at GCC 2024 and this roadmap by @nekrut: https://hackmd.io/@nekrut/HkinoEh8A

Improvements in https://github.com/galaxyproject/galaxy/pull/18762 further have further improved this interface as well.

## Overall Change to the component

| Before | After |
| ---- | ---- |
| ![invocation_view_improvements_before](https://github.com/user-attachments/assets/302c241d-a524-40a6-a7de-68fa4a3e8322) | ![invocation_view_improvements_after](https://github.com/user-attachments/assets/5ea3b227-e588-4aa2-8d3c-680cafe72291) |
| Steps show on the right, the progress bars are within the `Overview` tab | Only the currently active step shows below, leaving full width for graph, progress bars in the Header |

## Changes to an actively running workflow
Changes made to `WorkflowRunSuccess`:

| Before |
| ---- |
| ![workflow_run_success_before](https://github.com/user-attachments/assets/d2ee51dd-6762-4ec0-8f4b-a47d03cdb1b3) |

| After |
| ---- |
| ![workflow_run_success_after](https://github.com/user-attachments/assets/03450a43-691d-479a-8d87-1bd7557e8972) |

## Other Changes
Changes from the roadmap and other changes that improve the UX:

### - Steps moved to a separate tab, show clicked/active step below instead of job
This removes the steps from the overview and the graph is filling the center as the default view, and instead of clicking a step and then viewing a job below, the active/clicked step is shown below in its entirety:

https://github.com/user-attachments/assets/2748a3b4-0888-4b82-8cfe-fbf9eab1c1aa

### - Improved Workflow Invocation Step component
| Before | After |
| ---- | ---- |
| ![step_before](https://github.com/user-attachments/assets/973aee5d-dffe-4f6f-be63-01b0ce46f3b7) | ![step_after](https://github.com/user-attachments/assets/fe7c5dad-1c2f-4fbb-aee9-93ca4ab3bcec) |

### - Inputs shown in Graph
| Before | After |
| ---- | ---- |
| ![inputs_before](https://github.com/user-attachments/assets/58c27b54-3d08-4f0c-ac2d-41da06bd6056) | ![inputs_after](https://github.com/user-attachments/assets/a20d5cf3-197b-4073-8dc0-144e64d63851) |
| ![inputs_param_before](https://github.com/user-attachments/assets/56f9f71b-9ac6-4276-a8fa-464803f0c617) | ![inputs_param_after](https://github.com/user-attachments/assets/7f4cf667-9f26-4472-b9ed-b55a2575f1a1) |

### - Invocation Panel toggles out on `mouseleave`
The first commit tackles point A in the roadmap: List of invocations is now not visible continously, it goes away after you settle on one invocation and `mouseleave` the panel area:
![invocation_panel_toggled_mouseleave](https://github.com/user-attachments/assets/69371746-2e3d-4238-b94f-38e5df3e315f)

## How to test the changes?
(Select all options that apply)
- [ ] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

ahmedhamidawan committed:
Workflow Invocation view improvements

This first commit tackles point A in the roadmap:

List of invocations is now not visible continously, it goes away after you settle on one invocation and `mouseleave` the panel area.

Roadmap: https://hackmd.io/@nekrut/HkinoEh8A

ahmedhamidawan committed:
move invocation steps to a separate tab, show whole step below

- Instead of showing steps right next to the invocation graph, moves steps (back) into a separate tab
- To make this work, adds a `graphStepsByStoreId` to the `invocationStore` that helps access the graph steps in multiple components (TODO: improve how this is done)
- Instead of clicking on a node/step in the graph and showing the job for the step that the user clicks, now we show the whole expanded step below

ahmedhamidawan committed:
add workflow input indicator to `WorkflowInvocationStepHeader`

ahmedhamidawan committed:
decrease step header size in `InvocationGraph` to "sm"

add gap between step header buttons

ahmedhamidawan committed:
add a tab view for jobs under a step in invocation view

ahmedhamidawan committed:
use `openapi-fetch` for fetching job

ahmedhamidawan committed:
remove redundant props/emits for invocation steps

ahmedhamidawan committed:
convert `ProgressBar.vue` to composition API + ts

ahmedhamidawan committed:
do not include `Steps` tab for subworkflows

ahmedhamidawan committed:
add `getContentItemState` function

ahmedhamidawan committed:
add job duration, `WorkflowInvocationJob` component

remove `JobStepTabs.vue`

ahmedhamidawan committed:
check `activeNodeId` for `null` (it can be 0 as well)

ahmedhamidawan committed:
show workflow input value on graph

ahmedhamidawan committed:
move progress bars to header, shorten `WorkflowRunSuccess`

- The progress bars are in `WorkflowInvocationHeader` now, giving more space to the step header being visible below the graph
- `WorkflowRunSuccess` is a one line header (refactored), and it uses `GridInvocation` to show batch invocations.

ahmedhamidawan committed:
do not auto scroll to the clicked step

ahmedhamidawan committed:
add `force` prop to `WorkflowRunButton` for routing

ahmedhamidawan committed:
remove jest since buttons have moved to parent

ahmedhamidawan committed:
add a jest for `JobStep`

ahmedhamidawan committed:
adjust seleniums for changes in `JobStep` and invocations in general

ahmedhamidawan committed:
add a jest for `SwitchToHistoryLink` that tests current history case

For the current history, we show a title saying "This is your current history", therefore, a mock is added here for `historyStore.currentHistoryId` to always return `current-history-id` regardless of which history you mount. This is done so that we can test the case when the history is current, and not.

ahmedhamidawan committed:
do not auto-activate step on invocation view load

xref https://github.com/galaxyproject/galaxy/pull/18615#issuecomment-2457515715

ahmedhamidawan committed:
on a repetition of the step clicked, scroll to the step

ahmedhamidawan committed:
fix failing `WorkflowInvocationHeader` jest

- Adjust props
- Remove the return to invocation list jest, we will be removing that button soon

mvdbeek commented on client/src/components/WorkflowInvocationState/WorkflowInvocationJob.vue:
```diff
@@ -0,0 +1,38 @@
+        <div class="d-flex justify-content-between pb-1 pr-1">
+            <i>
+                <FontAwesomeIcon :icon="faRegClock" class="mr-1" />run
+                <UtcDate :date="props.job.create_time" mode="elapsed" />
```
This isn't the duration you would think it is, it's the entire time from scheduling to last to last update. Can you remove this ? Once the job is terminal you could get this from the wallclock job metric, but I would include that directly in the job info page.

ahmedhamidawan replied:
So, this is the `create_time`; I thought it is the time that the job was created?
In this `<i>` we show Job was run 18 minutes ago (e.g.)

Are you still talking about this being the _entire time from scheduling to last to last update_?

Isn't that the time in the next `<i>`, conditionally rendered for `v-if="!NON_TERMINAL_STATES.includes(props.job.state)"`?

which itself is calculated as [galaxy/client/src/components/JobInformation/utilities.ts](https://github.com/galaxyproject/galaxy/blob/8f6c95536f8e484f85a736436f93d9cc4980a27a/client/src/components/JobInformation/utilities.ts#L1-L7):
https://github.com/galaxyproject/galaxy/blob/8f6c95536f8e484f85a736436f93d9cc4980a27a/client/src/components/JobInformation/utilities.ts#L1-L7 ?

For that one, I can look at the wallclock job metric instead?


Here are the two different times shown here:
![image](https://github.com/user-attachments/assets/bfe22413-f03e-4290-ac7b-c5678aa1943a)

mvdbeek replied:
This is when the job was scheduled and first written to the database, that has nothing to do with what you'd call duration in a realistic production setup. duration would be wallclock time.

ahmedhamidawan replied:
I see.
 So should we remove either of these times from here?
If not, I understand the (terminal) one on the right is incorrect, and so do we still keep the one on the left?
And do we then place the job metric wallclock time here?

ahmedhamidawan replied:
So, I've removed both times from this view, and instead just simply used the `JobDetails` component that in turn uses the existing `JobMetrics` component in this latest commit https://github.com/galaxyproject/galaxy/pull/18615/commits/0c50efcb46763dd24c7e78db0c6d7c7967daffb9

https://github.com/user-attachments/assets/9aa39a53-6f78-40c4-8af4-129f1263bedd

Let me know if this works, since this is just a reuse of the existing job details view we see when we click on a `ContentItem`'s info option.

ahmedhamidawan committed:
show `JobMetrics` in workflow invocation job by just using `JobDetails`

No need for the `WorkflowInvocationJob` component added in this PR because `JobDetails` has all we need.
